redbeams/src/main/resources/schema/app/20190521145814_CB-1364_Database_server_uniqueness_constraint.sql

```diff
@@ -0,0 +1,24 @@
+-- // CB-1364 Database server uniqueness constraint
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE ONLY databaseserverconfig
+    DROP CONSTRAINT IF EXISTS uk_databaseserverconfig_deletiondate_workspace;
+
+-- PostgreSQL unique constraints always treat NULL values as unequal
+-- Therefore, use different constraints depending on value of deletiontimestamp
+
+CREATE UNIQUE INDEX uk_databaseserverconfig_deletionTimestamp_environment
+    ON databaseserverconfig (name, deletionTimestamp, environmentid)
```

Comment:
Two things. 1) Is this the best order for the fields? 2) You could probably use a single index where instead of deletionTimestamp you use COALESCE and a marker value.

Reply:
You clearly know more about indexing than I do, so maybe I should reorder the fields. Are they checked in that order?

I got the idea to create these "partial indices" from here: https://stackoverflow.com/questions/8289100/create-unique-constraint-with-null-columns and the second answer suggests using COALESCE. -1 would work as a marker value. Again, I'll defer to you - is that better?

Reply:
In general, the fewer indices, the better for insert performance. But I don't know how that would trade off with the additional cost of a function-based index. On the other issue about column order, in the absence of tailoring the column order to match a specific query, you usually want to reflect any external natural hierarchy in the column order. We're enforcing uniqueness of name within each environment, so it would be normal to put the environment id column first in the index. We're only using deletion timestamp to distinguish when a single server has been deleted and readded, so that would come last. I would put them in the order environmentId, name, deletionTimestamp, and my inclination would be to use COALESCE unless the stackoverflow article had specific reasons discouraging it.

Reply:
I'll reorder the fields as you suggest.

A comment in the SO question says that maintenance cost and storage are about the same vs a single index, and performance is better when a query doesn't cross over both of them (which would be common here). Basically, it seems like a wash, so because of that and because I'm lazy, I'll leave them as partial indices for now.